File: typescript/src/ancilis/engine/evaluators/gov01-policy.ts
/** GOV-01: Governance Policy evaluator. */

import type { Action } from "../action.js";
import type { ControlResult } from "../result.js";
import type { ResolvedConfig } from "../../config/index.js";
import type { ControlEvaluator } from "./base.js";

export class GOV01PolicyEvaluator implements ControlEvaluator {
  controlId = "GOV-01";
  controlName = "Governance Policy";

  evaluate(_action: Action, config: ResolvedConfig): ControlResult {
    const start = performance.now();

    const fieldsPresent: string[] = [];
    const fieldsMissing: string[] = [];

    // 1. agent_name must be set (non-empty)
    if (config.agentName?.trim()) {
      fieldsPresent.push("agent_name");
    } else {
      fieldsMissing.push("agent_name");
    }

    // 2. mode must be explicitly set
    if (config.mode?.trim()) {
      fieldsPresent.push("mode");
    } else {
      fieldsMissing.push("mode");
    }

    // 3. data_classifications must have at least one entry
    if (config.dataClassifications.size > 0) {
      fieldsPresent.push("data_classifications");
    } else {
      fieldsMissing.push("data_classifications");
    }

    // 4. scope constraints — tools_allowed or tools_blocked must be set
    const hasScope = config.toolsAllowed.length > 0 || config.toolsBlocked.length > 0;
    if (hasScope) {
      fieldsPresent.push("scope_constraints");
    } else {
      fieldsMissing.push("scope_constraints");
    }

    const presentCount = fieldsPresent.length;
    const total = fieldsPresent.length + fieldsMissing.length;

    let completeness: string;
    if (presentCount === total) {
      completeness = "complete";
    } else if (presentCount >= 2) {
      completeness = "partial";
    } else {
      completeness = "insufficient";
    }

    const evidence: Record<string, unknown> = {
      policy_completeness: completeness,
      fields_present: fieldsPresent,
      fields_missing: fieldsMissing,
    };

    const durationMs = performance.now() - start;

    if (completeness === "complete") {
      return {
        controlId: this.controlId,
        controlName: this.controlName,
        result: "PASS",
        detail: "Complete governance policy: all required fields are configured.",
        evidenceData: evidence,
        durationMs,
      };
    }

    if (completeness === "partial") {
      return {
        controlId: this.controlId,
        controlName: this.controlName,
        result: "FLAG",
        detail: `Partial governance policy: missing ${fieldsMissing.join(", ")}.`,
        evidenceData: evidence,
        durationMs,
      };
    }

    return {
      controlId: this.controlId,
      controlName: this.controlName,
      result: "FAIL",
      detail: `Insufficient governance policy: ${fieldsMissing.length} of ${total} required fields are missing.`,
      evidenceData: evidence,
      durationMs,
    };
  }
}

File: python/src/ancilis/engine/evaluators/gov01_policy.py
"""GOV-01: Governance Policy evaluator."""

from __future__ import annotations

import time
from typing import Any

from ancilis.config import ResolvedConfig
from ancilis.engine.action import Action
from ancilis.engine.result import ControlResult


class GOV01PolicyEvaluator:
    control_id = "GOV-01"
    control_name = "Governance Policy"

    def evaluate(self, action: Action, config: ResolvedConfig) -> ControlResult:
        start = time.perf_counter()

        fields_present: list[str] = []
        fields_missing: list[str] = []

        # 1. agent_name must be set (non-empty)
        if config.agent_name and config.agent_name.strip():
            fields_present.append("agent_name")
        else:
            fields_missing.append("agent_name")

        # 2. mode must be explicitly set
        if config.mode and config.mode.strip():
            fields_present.append("mode")
        else:
            fields_missing.append("mode")

        # 3. data_classifications must have at least one entry
        if config.data_classifications:
            fields_present.append("data_classifications")
        else:
            fields_missing.append("data_classifications")

        # 4. scope constraints — tools_allowed or tools_blocked must be set
        has_scope = bool(config.tools_allowed or config.tools_blocked)
        if has_scope:
            fields_present.append("scope_constraints")
        else:
            fields_missing.append("scope_constraints")

        total = len(fields_present) + len(fields_missing)
        present_count = len(fields_present)

        if present_count == total:
            completeness = "complete"
        elif present_count >= 2:
            completeness = "partial"
        else:
            completeness = "insufficient"

        evidence: dict[str, Any] = {
            "policy_completeness": completeness,
            "fields_present": fields_present,
            "fields_missing": fields_missing,
        }

        duration_ms = (time.perf_counter() - start) * 1000

        if completeness == "complete":
            return ControlResult(
                control_id=self.control_id,
                control_name=self.control_name,
                result="PASS",
                detail="Complete governance policy: all required fields are configured.",
                evidence_data=evidence,
                duration_ms=duration_ms,
            )

        if completeness == "partial":
            return ControlResult(
                control_id=self.control_id,
                control_name=self.control_name,
                result="FLAG",
                detail=f"Partial governance policy: missing {', '.join(fields_missing)}.",
                evidence_data=evidence,
                duration_ms=duration_ms,
            )

        return ControlResult(
            control_id=self.control_id,
            control_name=self.control_name,
            result="FAIL",
            detail=f"Insufficient governance policy: {len(fields_missing)} of {total} required fields are missing.",
            evidence_data=evidence,
            duration_ms=duration_ms,
        )
